ocfweb/docs/docs/staff/backend/kubernetes.md

```diff
@@ -0,0 +1,204 @@
+control plane, performs health checks, and recovers from pod failures should
+they occur.  Workers also run an instance of `kube-proxy`, which forwards
+control plane traffic to the correct `kubelet`.
+
+### Pods
+
+In the Kubernetes world, pods are the smallest computing unit. A pod is made up
+of one or more containers. The difference between a pod and a standalone
+container is best illustrated by an example. Consider [ocfweb][ocfweb]; it is
+composed of several containers—the web container, static container, and worker
+container.  In Kubernetes, together these three containers form one pod, and it
+is pods that can be scaled up or down. A failure in any of these containers
+indicates a failure in the entire pod. An astute reader might wonder: _if pods
+can be broken down into containers, how can pods possibly be the smallest
+unit?_ Do note that if one wished to deploy a singleton container, it would
+still need to be wrapped in the pod abstraction for Kubernetes to do anything
+with it.
+
+While pods are essential for understanding Kubernetes, when writing services we
+don't actually deal in pods but one further abstraction,
+[deployments][deployment], which create pods for us.
+
+## OCF Kubernetes Cluster Bootstrapping
+
+Since almost all OCF architecture is bootstapped using Puppet, it was necessary
+for us to do the same with Kubernetes. We rely on the
+[puppetlabs-kubernetes][kubernetes-module] module to handle initial
+bootstrapping and bolt OCF specific configurations on top of it.
+`puppetlabs-kubernetes` performs two crucial tasks:
+
+1. Installs `etcd`, `kubelet`, `kube-proxy`, and `kube-dns`, initializes the
+   cluster, and applies a networking backend.
+2. Generates the required [PKI for Kubernetes and etcd][kubernetes-pki].
+
+Do note that `puppetlabs-kubernetes` is still very much a work in progress. If
+you notice an issue in the module you are encouraged to write a patch and send
+it upstream.
+
+### Kubernetes PKI
+
+All the private keys and certs for the PKI are in the puppet private share, in
+`/opt/puppet/shares/private/kubernetes`. We won't go into detail of everything
+contained there, but Kubernetes and `etcd` communication is authenticated using
+client certificates. All the necessary items for workers are included in
+`os/Debian.yaml`, although adding a new master to the cluster requires
+re-running [kubetool][puppetlabs-kubetool] to generate new `etcd server` and
+`etcd peer` certs.
+
+### OCF Kubernetes Configuration
+
+Currently, the OCF has three Kubernetes masters: (1) `deadlock`, (2) `coup`,
+and (3) `autocrat`. A Container Networking Interface (`cni`) is the last piece
+required for a working cluster. The `cni`'s purpose is to faciltate intra-pod
+communication. `puppetlabs-kubernetes` supports two choices: `weave` and
+`flannel`. Both solutions work out-the-box, and we've had success with
+`flannel` thus far so we've stuck with it.
+
+### Getting traffic into the cluster
```

Comment:
I'd say maybe this is a level 2 header, since it's not really under bootstrapping any more, more just regular operation, and then the level 3 stuff under it makes sense since it fits under this header.